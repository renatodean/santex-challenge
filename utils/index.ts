import { Item } from "../types/product";

/**
 * From a given product it will return the price as a formated string
 * @param product
 * @param quantity
 * @returns string
 */
export const getPrice = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return ((productPrice * (quantity || 1)) / 100).toFixed(2);
};

/**
 * From a given product it will return the price as number
 * @param product
 * @param quantity
 * @returns number
 */
export const getPriceNumber = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return (productPrice * (quantity || 1)) / 100;
};

/**
 * From a given query from router it will return the query value
 * @param query
 * @returns string
 */
export const getQueryValue = (query: any) => {
  if (Array.isArray(query)) {
    return query[0];
  } else {
    return query || "";
  }
};
