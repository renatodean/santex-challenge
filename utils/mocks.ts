export const mockedProducts = {
  products: {
    items: [
      {
        __typename: "Product",
        id: "1",
        name: "Laptop",
        slug: "laptop",
        description:
          "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
        featuredAsset: {
          __typename: "Asset",
          id: "1",
          name: "derick-david-409858-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 129900 },
          { __typename: "ProductVariant", price: 139900 },
          { __typename: "ProductVariant", price: 219900 },
          { __typename: "ProductVariant", price: 229900 },
        ],
      },
      {
        __typename: "Product",
        id: "2",
        name: "Tablet",
        slug: "tablet",
        description:
          'If the computer were invented today, what would it look like? It would be powerful enough for any task. So mobile you could take it everywhere. And so intuitive you could use it any way you wanted — with touch, a keyboard, or even a pencil. In other words, it wouldn’t really be a "computer." It would be Tablet.',
        featuredAsset: {
          __typename: "Asset",
          id: "2",
          name: "kelly-sikkema-685291-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/5a/kelly-sikkema-685291-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/b8/kelly-sikkema-685291-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 32900 },
          { __typename: "ProductVariant", price: 44500 },
        ],
      },
    ],
  },
};

export const mockedEmptyProducts = { products: { items: [] } };

export const mockedProduct = {
  product: {
    __typename: "Product",
    id: "1",
    name: "Laptop",
    slug: "laptop",
    description:
      "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
    featuredAsset: {
      __typename: "Asset",
      id: "1",
      name: "derick-david-409858-unsplash.jpg",
      type: "IMAGE",
      source:
        "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
      preview:
        "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
    },
    variants: [
      { __typename: "ProductVariant", price: 129900 },
      { __typename: "ProductVariant", price: 139900 },
      { __typename: "ProductVariant", price: 219900 },
      { __typename: "ProductVariant", price: 229900 },
    ],
  },
};

export const mockedStorageState = {
  showSnackbar: true,
  products: [
    {
      product: mockedProduct.product,
      quantity: 3,
    },
  ],
};
