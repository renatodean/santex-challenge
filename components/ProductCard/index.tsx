import React from "react";
import styled from "@emotion/styled";
import { Item } from "../../types/product";
import Image from "next/image";
import { getPrice } from "../../utils";
import { useRouter } from "next/router";

const Container = styled.div`
  padding: 10px;
  margin: 10px;
  border: 0.5px solid rgba(0, 0, 0, 20%);
  border-radius: 8px;
  background-color: white;
  max-width: 200px;
  display: flex;
  flex-direction: column;
  cursor: pointer;
  text-align: center;
  &:hover {
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  }
`;

const ImgContainer = styled.div`
  position: relative;
  width: 180px;
  height: 180px;
`;

const ProductName = styled.span`
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
  max-width: 180px;
  border-top: 0.5px solid gray;
  padding-top: 10px;
  margin-top: 10px;
  margin-bottom: 4px;
`;

const ProductPrice = styled.span`
  font-size: 16px;
  line-height: 20px;
  font-weidth: 600;
  max-width: 180px;
`;

interface ProductCardProps {
  product: Item;
}

const ProductCard = ({ product }: ProductCardProps) => {
  const router = useRouter();

  return (
    <Container
      tabIndex={0}
      onClick={() => router.push(`/product/${product.slug}`)}
      onKeyPress={(event) => {
        if (event.key === "Enter") {
          router.push(`/product/${product.slug}`);
        }
      }}
    >
      <ImgContainer>
        <Image
          src={product.featuredAsset.source}
          priority
          alt={product.name}
          layout="fill"
          objectFit="contain"
        />
      </ImgContainer>
      <ProductName>{product.name}</ProductName>
      <ProductPrice>${getPrice(product)}</ProductPrice>
    </Container>
  );
};

export default ProductCard;
