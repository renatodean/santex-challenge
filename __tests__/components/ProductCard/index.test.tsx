import { render, screen, fireEvent } from "@testing-library/react";
import ProductCard from "../../../components/ProductCard";
import { mockedProduct } from "../../../utils/mocks";

const routerPushSpy = jest.fn();

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: "",
      asPath: "",
      push: routerPushSpy,
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

describe("ProductCard", () => {
  it("renders the product information", () => {
    render(<ProductCard product={mockedProduct.product} />);

    expect(
      screen.getByText(mockedProduct.product.name)
    ).toBeInTheDocument();
  });

  it("should go to a product when press enter", () => {
    render(<ProductCard product={mockedProduct.product} />);

    const productCard = screen.getByTestId("product-card");
    fireEvent.keyPress(productCard, {key: 'Enter', code: 'Enter', charCode: 13})

    expect(routerPushSpy).toHaveBeenCalled();
    expect(routerPushSpy).toHaveBeenCalledTimes(1);
    expect(routerPushSpy).toHaveBeenCalledWith("/product/laptop");
  });
});

describe("ProductCard snapshot", () => {
  it("renders product detail", () => {
    const { container } = render(
      <ProductCard product={mockedProduct.product} />
    );
    expect(container).toMatchSnapshot();
  });
});
