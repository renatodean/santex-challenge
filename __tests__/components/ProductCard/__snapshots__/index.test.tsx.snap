// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard snapshot renders product detail 1`] = `
<div>
  <div
    class="css-10s52hl"
    data-testid="product-card"
    tabindex="0"
  >
    <div
      class="css-1jqp1sw"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="Laptop"
          data-nimg="fill"
          decoding="async"
          sizes="100vw"
          src="/_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=3840&q=75 3840w"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
        />
      </span>
    </div>
    <span
      class="css-1o0g9xy"
    >
      Laptop
    </span>
    <span
      class="css-1v50op4"
    >
      $
      1299.00
    </span>
  </div>
</div>
`;
